MatrixSDK/Crypto/KeySharing/MXIncomingRoomKeyRequestManager.m

```diff
@@ -0,0 +1,288 @@
+    // cancellation before the request (and end up with an outstanding
+    // request which should have been cancelled.)
+    for (MXIncomingRoomKeyRequest *req in requests)
+    {
+        [self processReceivedRoomKeyRequest:req];
+    }
+    for (MXIncomingRoomKeyRequestCancellation *cancellation in cancellations)
+    {
+        [self processReceivedRoomKeyRequestCancellation:cancellation];
+    }
+}
+
+/**
+ Helper for processReceivedRoomKeyRequests.
+
+ @param req the request.
+ */
+- (void)processReceivedRoomKeyRequest:(MXIncomingRoomKeyRequest*)req
+{
+    NSString *userId = req.userId;
+    NSString *deviceId = req.deviceId;
+    NSString *requestId = req.requestId;
+
+    NSDictionary *body = req.requestBody;
+    NSString *roomId, *alg;
+
+    MXJSONModelSetString(roomId, body[@"room_id"]);
+    MXJSONModelSetString(alg, body[@"algorithm"]);
+
+    NSLog(@"[MXIncomingRoomKeyRequestManager] processReceivedRoomKeyRequest: m.room_key_request from %@:%@ for %@ / %@ (id %@)", userId, deviceId, roomId, body[@"session_id"], req.requestId);
+
+    if (![userId isEqualToString:crypto.matrixRestClient.credentials.userId])
+    {
+        // TODO: determine if we sent this device the keys already: in
+        // which case we can do so again.
+        NSLog(@"[MXIncomingRoomKeyRequestManager] Ignoring room key request from other user for now");
+        return;
+    }
+
+    // todo: should we queue up requests we don't yet have keys for,
+    // in case they turn up later?
+
+    // if we don't have a decryptor for this room/alg, we don't have
+    // the keys for the requested events, and can drop the requests.
+    id<MXDecrypting> decryptor = [crypto getRoomDecryptor:roomId algorithm:alg];
+    if (!decryptor)
+    {
+        NSLog(@"[MXIncomingRoomKeyRequestManager] room key request for unknown alg %@ in room %@", alg, roomId);
+        return;
+    }
+
+    if (![decryptor hasKeysForKeyRequest:req])
+    {
+        NSLog(@"[MXIncomingRoomKeyRequestManager] room key request for unknown session %@ / %@", roomId, body[@"session_id"]);
+        return;
+    }
+
+    // if the device is verified already, share the keys
+    MXDeviceInfo *device = [crypto.store deviceWithDeviceId:deviceId forUser:userId];
+    if (device && device.verified == MXDeviceVerified)
+    {
+        NSLog(@"[MXIncomingRoomKeyRequestManager] device is already verified: sharing keys");
+        [decryptor shareKeysWithDevice:req success:nil failure:nil];
+        return;
+    }
+
+    // check if we already have this request
+    if ([self pendingKeyRequest:requestId fromUser:userId andDevice:deviceId])
+    {
+        NSLog(@"[MXIncomingRoomKeyRequestManager] Already have this key request, ignoring");
+        return;
+    }
+
+    // Add it to pending key requests
+    [self addPendingKeyRequest:req];
+
+    // Broadcast the room key request
+        __weak typeof(self) weakSelf = self;
+    dispatch_async(dispatch_get_main_queue(), ^{
+        if (weakSelf)
```

Comment:
Why do you add here a sanity check on self?

Reply:
you are right. It would be better to make it on `crypto`.